liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+                                    }
+                                }
+                            } catch (DependencyResolutionException e) {
+                                throw new MojoExecutionException(e.getMessage(), e);
+                            }
+                            artifactPaths.addAll(addToClassPath);
+                        }
+
+                        if (!dependencyIds.isEmpty()) {
+                            runMojo("net.wasdev.wlp.maven.plugins:liberty-maven-plugin", "install-feature", serverName,
+                                    dependencyIds);
+                            dependencyIds.clear();
+                        }
+
+                        // update dependencies
+                        this.existingDependencies = dependencies;
+                        this.existingPom = modifiedPom;
+
+                        return true;
+                    } else {
+                        log.info("Unhandled change detected in pom.xml. Restart liberty:dev mode for it to take effect.");
+                    }
+                }
+
+            } catch (Exception e) {
+                log.debug("Could not recompile pom.xml", e);
+            }
+            return false;
+        }
+
+        @Override
+        public int countApplicationUpdatedMessages() {
+            int messageOccurrences = -1;
+            if (!(skipTests || skipITs)) {
+                try {
+                    File logFile = serverTask.getLogFile();
+                    String regexp = UPDATED_APP_MESSAGE_REGEXP + DevMojo.this.project.getArtifactId();
+                    messageOccurrences = serverTask.countStringOccurrencesInFile(regexp, logFile);
+                    log.debug("Message occurrences before compile: " + messageOccurrences);
+                } catch (Exception e) {
+                    log.debug("Failed to get message occurrences before compile", e);
+                }
+            }
+            return messageOccurrences;
+        }
+
+        @Override
+        public void runTests(boolean waitForApplicationUpdate, int messageOccurrences, ThreadPoolExecutor executor,
+                boolean forceSkipUTs) {
+            if (!skipTests) {
+                File logFile = serverTask.getLogFile();
+                String regexp = UPDATED_APP_MESSAGE_REGEXP + DevMojo.this.project.getArtifactId();    
+
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException e) {
+                    log.debug("Thread interrupted while waiting to start unit tests.", e);
+                }
+
+                // if queue size >= 1, it means a newer test has been queued so we
+                // should skip this and let that run instead
+                if (executor.getQueue().size() >= 1) {
+                    Runnable head = executor.getQueue().peek();
+                    boolean manualInvocation = ((TestJob) head).isManualInvocation();
```

Comment:
What is `TestJob` from?

Reply:
`TestJob` is from ci.common.  See https://github.com/WASdev/ci.common/blob/7483bd22ac844f7cb8d18c0eae7a9e4558c3f658/src/main/java/net/wasdev/wlp/common/plugins/util/DevUtil.java#L949 .  We need to check whether the test request was manually invoked or run automatically after code changes, then print the proper message.